src/ess/reduce/nexus/types.py

```diff
@@ -25,80 +26,223 @@
 
-AnyRunFilename = FilePath | NeXusFile | NeXusGroup
+class NeXusMonitorName(sciline.Scope[MonitorType, str], str):
+    """Name of a monitor in a NeXus file."""
+
+
+class NeXusDetector(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Full raw data from a NeXus detector."""
+
+
+class NeXusMonitor(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataGroup], sc.DataGroup
+):
+    """Full raw data from a NeXus monitor."""
+
+
+class NeXusSample(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Raw data from a NeXus sample."""
+
+
+class NeXusSource(sciline.Scope[RunType, sc.DataGroup], sc.DataGroup):
+    """Raw data from a NeXus source."""
+
+
+class NeXusDetectorData(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Data array loaded from an NXevent_data or NXdata group within an NXdetector."""
+
+
+class NeXusMonitorData(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Data array loaded from an NXevent_data or NXdata group within an NXmonitor."""
+
+
+class SourcePosition(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Position of the neutron source."""
+
+
+class SamplePosition(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Position of the sample."""
+
+
+class DetectorPositionOffset(sciline.Scope[RunType, sc.Variable], sc.Variable):
+    """Offset for the detector position, added to base position."""
 
-PreopenNeXusFile = NewType('PreopenNeXusFile', bool)
-"""Whether to preopen NeXus files before passing them to the rest of the workflow."""
+
+class MonitorPositionOffset(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.Variable], sc.Variable
+):
+    """Offset for the monitor position, added to base position."""
+
+
+class CalibratedDetector(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Calibrated data from a detector."""
+
+
+class CalibratedMonitor(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Calibrated data from a monitor."""
+
+
+class DetectorData(sciline.Scope[RunType, sc.DataArray], sc.DataArray):
+    """Calibrated detector merged with neutron event or histogram data."""
+
+
+class MonitorData(
+    sciline.ScopeTwoParams[RunType, MonitorType, sc.DataArray], sc.DataArray
+):
+    """Calibrated monitor merged with neutron event or histogram data."""
+
+
+class Filename(sciline.Scope[RunType, Path], Path): ...
+
+
+@dataclass
+class PulseSelection(Generic[RunType]):
+    """Range of neutron pulses to load from NXevent_data or NXdata groups."""
+
+    value: snx.typing.ScippIndex | slice
+
+    def to_snx_selection(self, *, for_events: bool) -> snx.typing.ScippIndex:
+        if self.value == slice(None, None):
+            return ()
+        if isinstance(self.value, slice):
+            if for_events:
+                return {'event_time_zero': self.value}
+            return {'time': self.value}
+        return self.value
```

Comment:
Duplicate of `_nexus_loader._to_snx_selection`. Why did you change the loader to take a raw selection instead of a `PulseSelection` object?

Reply:
Because `PulseSelection` is now generic, so that didn't work.

Thought I removed this (I was moving it back and forth), good catch.